src/Microsoft.ML.Data/Transforms/KeyToVectorTransform.cs

```diff
@@ -69,530 +75,776 @@ public bool TryUnparse(StringBuilder sb)
-                    }
+        public KeyToVectorEstimator(IHostEnvironment env, KeyToVectorTransform transformer)
+            : base(Contracts.CheckRef(env, nameof(env)).Register(nameof(KeyToVectorEstimator)), transformer)
+        {
+        }
 
-                    var values = dst.Values;
-                    var indices = dst.Indices;
-                    if (Utils.Size(values) < 1)
-                        values = new Float[1];
-                    if (Utils.Size(indices) < 1)
-                        indices = new int[1];
-                    values[0] = 1;
-                    indices[0] = (int)src - 1;
+        public override SchemaShape GetOutputSchema(SchemaShape inputSchema)
+        {
+            Host.CheckValue(inputSchema, nameof(inputSchema));
+            var result = inputSchema.Columns.ToDictionary(x => x.Name);
+            foreach (var colInfo in Transformer.Columns)
+            {
+                if (!inputSchema.TryFindColumn(colInfo.Input, out var col))
+                    throw Host.ExceptSchemaMismatch(nameof(inputSchema), "input", colInfo.Input);
+                if ((col.ItemType.ItemType.RawKind == default) || !(col.ItemType.IsVector || col.ItemType.IsPrimitive))
+                    throw Host.ExceptSchemaMismatch(nameof(inputSchema), "input", colInfo.Input);
+
+                var metadata = new List<SchemaShape.Column>();
+                if (col.Metadata.TryFindColumn(MetadataUtils.Kinds.KeyValues, out var keyMeta))
+                    if (col.Kind != SchemaShape.Column.VectorKind.VariableVector && keyMeta.ItemType.IsText)
+                        metadata.Add(new SchemaShape.Column(MetadataUtils.Kinds.SlotNames, SchemaShape.Column.VectorKind.Vector, keyMeta.ItemType, false));
+                if (!colInfo.Bag && (col.Kind == SchemaShape.Column.VectorKind.Scalar || col.Kind == SchemaShape.Column.VectorKind.Vector))
+                    metadata.Add(new SchemaShape.Column(MetadataUtils.Kinds.CategoricalSlotRanges, SchemaShape.Column.VectorKind.Vector, NumberType.I4, false));
+                if (!colInfo.Bag || (col.Kind == SchemaShape.Column.VectorKind.Scalar))
+                    metadata.Add(new SchemaShape.Column(MetadataUtils.Kinds.IsNormalized, SchemaShape.Column.VectorKind.Scalar, BoolType.Instance, false));
+
+                result[colInfo.Output] = new SchemaShape.Column(colInfo.Output, SchemaShape.Column.VectorKind.Vector, NumberType.R4, false, new SchemaShape(metadata));
+            }
 
-                    dst = new VBuffer<Float>(size, 1, values, indices);
-                };
+            return new SchemaShape(result.Values);
         }
+    }
 
-        /// <summary>
-        /// This is for the bagging case - vector input and outputs should be added.
-        /// </summary>
-        private ValueGetter<VBuffer<Float>> MakeGetterBag(IRow input, int iinfo)
+    /// <summary>
+    /// Extension methods for the static-pipeline over <see cref="PipelineColumn"/> objects.
```

Comment:
>static-pipeline over <see cref="PipelineColumn"/> objects [](start = 34, length = 57)

I I think you meant static pipelines over the key-to-vector transform, right?

Reply:
I just copied this line from other code in hope what this is common pattern. Is it not the case?

---
In reply to: [216801837](https://github.com/dotnet/machinelearning/pull/858#discussion_r216801837) [](ancestors = 216801837)

Reply:
If it is I or someone erred.

---
In reply to: [216812383](https://github.com/dotnet/machinelearning/pull/858#discussion_r216812383) [](ancestors = 216812383,216801837)